File: examples/src/main/java/com/edgedb/examples/BasicQueryFunctions.java
package com.edgedb.examples;

import com.edgedb.driver.EdgeDBClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletionStage;

public final class BasicQueryFunctions implements Example {
    private static final Logger logger = LoggerFactory.getLogger(BasicQueryFunctions.class);

    @Override
    public CompletionStage<Void> run(EdgeDBClient client) {
        return client.query(String.class, "select 'Hello, Java!'")
                .thenAccept(result -> logger.info("query result: {}", result))
                .thenCompose(v -> client.querySingle(String.class, "select 'Hello, Java!'"))
                .thenAccept(result -> logger.info("querySingle result: {}", result))
                .thenCompose(v -> client.queryRequiredSingle(String.class, "select 'Hello, Java!'"))
                .thenAccept(result -> logger.info("queryRequiredSingle result: {}", result));
    }
}

File: examples/scala-examples/src/main/scala/BasicQueryFunctions.scala
package com.edgedb.examples
import com.edgedb.driver.EdgeDBClient
import org.slf4j.LoggerFactory

import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.FutureConverters.*
class BasicQueryFunctions extends Example:
  private val logger = LoggerFactory.getLogger(classOf[BasicQueryFunctions])

  override def run(client: EdgeDBClient)(implicit context: ExecutionContext): Future[Unit] = {
    for {
      queryResult <- client.query(classOf[String], "SELECT 'Hello, Scala!'").asScala
      querySingleResult <- client.querySingle(classOf[String], "SELECT 'Hello, Scala!'").asScala
      queryRequiredSingleResult <- client.queryRequiredSingle(classOf[String], "SELECT 'Hello, Scala!'").asScala
    }
    yield (queryResult, querySingleResult, queryRequiredSingleResult) {
      logger.info("'query' method result: {}", queryResult)
      logger.info("'querySingle' method result: {}", querySingleResult)
      logger.info("'queryRequiredSingle' method result: {}", queryRequiredSingleResult)
      0
    }
  }